client/shared/src/test/scala/org/http4s/client/ClientRouteTestBattery.scala

```diff
@@ -129,6 +129,51 @@ abstract class ClientRouteTestBattery(name: String) extends Http4sSuite with Htt
     }
   }
 
+  GetRoutes.getPaths.get(GetRoutes.SimplePath).foreach { expected =>
+    val path = GetRoutes.SimplePath
+
+    test(s"$name Execute GET withEntity $path") {
+      serverClient().flatMap { case (server, client) =>
+        val address = server().addresses.head
+        val req =
+          Request[IO](uri = Uri.fromString(s"http://$address$path").yolo).withEntity("entity")
+        client()
+          .run(req)
+          .use(resp => expected.flatMap(checkResponse(resp, _)))
+          .assert
+      }
+    }
+
+    // test(s"$name Execute GET withEntity then withHeaders $path") {
+    //   serverClient().flatMap { case (server, client) =>
+    //     val address = server().addresses.head
+    //     val req =
+    //       Request[IO](uri = Uri.fromString(s"http://$address$path").yolo)
+    //         .withEntity("entity")
+    //         .withHeaders(headers.`Content-Type`(MediaType.text.plain))
+    //     client()
+    //       .run(req)
+    //       .use(resp => expected.flatMap(checkResponse(resp, _)))
+    //       .assert
+    //   }
+    // }
```

Comment:
This addresses 1 [above](https://github.com/http4s/http4s/pull/6117/files/0a043b23bd9691c3e0a9ab26baba3c03236c1a93#r826395887), and now the ember-client tests pass for me:

```suggestion
    test(s"$name Execute GET withEntity then putHeaders $path") {
      serverClient().flatMap { case (server, client) =>
        val address = server().addresses.head
        val req =
          Request[IO](uri = Uri.fromString(s"http://$address$path").yolo)
            .withEntity("entity")
            .putHeaders(headers.`Content-Type`(MediaType.text.plain))
        client()
          .run(req)
          .use(resp => expected.flatMap(checkResponse(resp, _)))
          .assert
      }
    }
```

I suspect you're right on point 2, which is worth a separate issue.

Reply:
Right, the point I tried to make with these tests is that there are surprises in the use of `withHeaders`, i.e.
the request that's created via `req.withEntity(..).withHeaders(..)` is different than the request created via `req.withHeaders(...).withEntity(...)`. I think there would be less surprises if the semantics would be like I described in point 1 [here](https://github.com/http4s/http4s/pull/6117#discussion_r826395887) (but since it changes current behaviour and I'm not sure whether users rely on this behaviour somehow, I don't know if it should be changed at this stage).

Note that it's not just `withEntity(..).withHeaders(..)` that's doesn't create content length header, this test also fails for Ember also because the constructed request doesn't have content length:

```
    test(s"$name Execute GET with Request(uri, headers, entity) $path".only) {
      serverClient().flatMap { case (server, client) =>
        val address = server().addresses.head
        val req =
          Request[IO](
            uri = Uri.fromString(s"http://$address$path").yolo, 
            headers = Headers(headers.`Content-Type`(MediaType.text.plain)),
            entity = EntityEncoder.stringEncoder.toEntity("entity"))
        client()
          .run(req)
          .use(resp => expected.flatMap(checkResponse(resp, _)))
          .assert
      }
    }
  }
```